### DISCUSSION TITLE
Getting ill (?) formatted config back from webapp

### DISCUSSION THREAD

**Discussion by aerofeev2k** [oracle/opengrok#3935] [Q&A] [answered]
On several occasions I've seen config in webapp getting somehow messed up and then returned to me via REST with fields like below:

```
   <void method="put">
    <string>PROJECT</string>
    <object class="org.opengrok.indexer.configuration.Project">
     <void property="historyEnabled">
      <object idref="Boolean0"/>
     </void>
     <void property="indexed">
      <object idref="Boolean0"/>
     </void>
     <void property="name">
      <string>PROJECT</string>
     </void>
     <void property="path">
      <string>/PROJECT</string>
     </void>
     <void property="tabSize">
      <object idref="Integer0"/>
     </void>
    </object>
   </void>

   <void property="rebuildThreadPoolSizeInNcpuPercent">
    <void idref="Configuration0">
     <void id="Integer2" property="maxRevisionThreadCount"/>
     <void property="tabSize">
      <object idref="Integer0"/>
     </void>
     <void property="userPage">
      <string>https://opengrok/user/</string>
     </void>
    </void>
    <object idref="Integer2"/>
   </void>
```

Initially (after a restart and reload) it all looks good, i.e. w/o these idref, Integer0, Integer2 or Boolean0, but after some time and some activity (I'm not quite sure, which one specifically triggers this, they're all like marking projects indexed, getting/setting config and kicking off Suggesters), I start getting the above.

This doesn't look like a correct format, right? What could be the reason?

**Comment by aerofeev2k** [comment]
Actually, just querying config _a few times_ is enough to trigger the above output:

`curl -s -X GET "$URL/configuration"`

So, if I restart Tomcat, wait for it to load and then run the above command, I get this:

```
...
    <void method="add">
     <string>type</string>
    </void>
    <void method="add">
     <string>full</string>
    </void>
   </void>
   <void property="rebuildThreadPoolSizeInNcpuPercent">
    <int>40</int>
   </void>
  </void>
  <void property="tabSize">
   <int>4</int>
  </void>
  <void property="userPage">
   <string>https://server/usermanagement/user/</string>
  </void>
 </object>
</java>
```

But after trying it 6 more times I get this:

```
...
    <void method="add">
     <string>full</string>
    </void>
   </void>
   <void property="rebuildThreadPoolSizeInNcpuPercent">
    <void idref="Configuration0">
     <void id="Integer2" property="maxRevisionThreadCount"/>
     <void property="tabSize">
      <object idref="Integer0"/>
     </void>
     <void property="userPage">
      <string>https://server/usermanagement/user/</string>
     </void>
    </void>
    <object idref="Integer2"/>
   </void>
  </void>
 </object>
</java>
```

**Comment by vladak** [comment, marked as answer]
As long as the configuration can be de-serialized the format is fine. It's the way XML serializer works in Java. The `idref` replaces object containment with object reference.

https://docs.oracle.com/javase/7/docs/api/java/beans/XMLEncoder.html says:
>  XML's standard "id" and "idref" attributes are used to make references to previous expressions - so as to deal with circularities in the object graph. 

Also see https://docs.oracle.com/javase/8/docs/api/javax/xml/bind/annotation/XmlIDREF.html

This is just one of the cons of the XML serialization. #2329 tracks the replacement of the serialization scheme.

**Reply by aerofeev2k** [reply]
I may be missing something, but how is it fine if I don't see the actual values for rebuildThreadPoolSizeInNcpuPercent, indexed or historyEnabled in the XML?

Can I persist config in this format to /etc/opengrok/configuration.xml, then restart Tomcat and expect it to read it?

**Reply by vladak** [reply]
Unless there is a bug in how Java (de)serializes the objects, it should be possible to arrive to the same state after the webapp is restarted,
